### DISCUSSION TITLE
What is RequextContext's next parameter?

### DISCUSSION THREAD

**Discussion by danielniccoli** [mikro-orm/mikro-orm#5367] [Q&A] [answered]
I'm trying to get MikroORM running with SvelteKit. I found a discussion about this here: https://github.com/mikro-orm/mikro-orm/discussions/4007

What I'm trying to understand is the `next` parameter that `RequextContext` expects. The [docs](https://mikro-orm.io/api/next/core/class/RequestContext#create) are not explaining this.

From the discussion, it just looks like a response object ...

```js
const next = async () => {
    response = await resolve(event);
    return response;
  };
```

... that maybe even could be simplified as

```js
const next = await resolve(event);
```

But if that's true, why would Mikro ORM need a response object in the first place?

**Comment by B4nan** [comment]
Its a callback, not an object.

> why would Mikro ORM need a response object in the first place?

It doesn't, you need to wrap your request handler in a context - you need to declare where the request lifecycle starts and ends (beginning and ending of the `next` callback).

**Reply by danielniccoli** [reply]
> It doesn't, you need to wrap your request handler in a context - you need to declare where the request lifecycle starts and ends (beginning and ending of the next callback).

I don't understand what you're saying.

**Reply by B4nan** [reply]
Do you understand what is a middleware?

**Reply by B4nan** [reply]
And have you read the docs around this? 

https://mikro-orm.io/docs/identity-map

**Reply by B4nan** [reply]
Did a quick search and you'll probably need a svetle node adapter for this, svetle itself is not handling any lifecycle for you.

https://kit.svelte.dev/docs/adapter-node
https://stackoverflow.com/questions/73090527/how-can-i-make-middleware-with-svelte

**Reply by B4nan** [reply]
Also found this discussion, the solution seems pretty similar to what you were talking about 

https://github.com/mikro-orm/mikro-orm/discussions/4007

**Reply by danielniccoli** [reply]
> Do you understand what is a middleware?

I understand what is a middleware.

> And have you read the docs around this?

I have read the docs, but they don't really explain anything to someone not knowing what an Identity Map is. 
Searching for Identity Map, leads me to a very brief wiki article and after reading that, I'm not sure if I need to understand all of that.

I have experience with Django ORM, Prisma and a little of .NET Entity Framework and haven't had such trouble understanding the docs, as I have with Mikro ORM. I could implement or use all of these frameworks without having to know so much in-depth. For example, Django ORM uses lazy loading (which is similar to Identity Map, as far as I can tell with my limited knowledge) and I can start using it without having to know how it internally works.

> Did a quick search and you'll probably need a svetle node adapter for this, svetle itself is not handling any lifecycle for you.

Okay, unfortunately this is the point where I'll have to start looking for another ORM, because that's just an additional requirement where I'm pretty sure I won't find an answer to.

> Also found this discussion, the solution seems pretty similar to what you were talking about

That's the one I linked, and where I still wanted to understand a little about the `next` parameter.


Thanks for your help.

**Reply by B4nan** [reply]
I can try to create a demo svetle app for you, but if you want a tool you already understand, or you want something right ahead, better to look elsewhere 🤷

**Reply by danielniccoli** [reply]
Ok, let's do it together. I've already setup a sample repository where I could get the basics working: https://github.com/danielniccoli/mikroorm-svelte

In `db.js` the Mikro ORM is configured.

In `hooks.server.js` there are two places where `RequestContext.create` can be placed. Outside the handle function, which runs the code exactly once, or inside the handle function that runs on every request.

If you visit `/` you should see either ` Database connected = true` or ` Database connected = false`. The value is retrieved in `+page.server.js` via `await orm.checkConnection()` and then rendered through `+page.svelte`.

If you could help me understand why the `next` parameter is required, we can figure out where that part should live within `hooks.server.js` and find a way to set it up correctly.

**Reply by B4nan** [reply, marked as answer]
You want something like this:

```ts
export async function handle({ event, resolve }) {
    return RequestContext.create(em, async () => {
        return await resolve(event)
    })
}
```

The resolve call needs to happen inside the callback. [Sent you a PR](https://github.com/danielniccoli/mikroorm-svelte/pull/1) that also changes few other things which I had to do to be able to actually test this. 

Namely:
- added CLI and refactored the config so it can be shared with the CLI
- used the EM inside your page handler, that is what would throw if you wouldn't have the context set up properly, as using global context is disallowed. if you try to remove the request context handling, you will see yourself
- added logging of the `em.id` to showcase what this is about - every request has its own, this is important to allow parallel request processing

**Reply by danielniccoli** [reply]
Okay, I finally understood how that works and why the callback is required.

I could also do the following to create a context outside of `resolve(event)`, if I'd ever need it:

```js
import { AsyncLocalStorage } from "async_hooks";
import { RequestContext } from "@mikro-orm/core"
import { orm } from "$lib/db"

/** @type {import('@sveltejs/kit').Handle} */
export async function handle({ event, resolve }) {
    // Create a Mikro-ORM context that exists through the lifetime of the web request.
    let ctx = RequestContext.createContext(orm.em)
    // Store the Mikro-ORM context into the store for the entire execution of this async function
    RequestContext.storage.enterWith(ctx)

    const response = await resolve(event);
    return response
}
```

Works the same.
